File: src/domain/analysis.ts
import type { Photo, PhotoFingerprint } from "./types.ts";

export interface PhotoGroup {
  id: string;
  ids: string[];
  recommended: string;
  kind: "similar" | "duplicate";
}
export interface LibraryAnalysis {
  status: "idle" | "scanning" | "complete" | "unavailable" | "error";
  processed: number;
  total: number;
  unavailable: number;
  groups: PhotoGroup[];
  error: string;
}
export const emptyAnalysis: LibraryAnalysis = {
  status: "idle",
  processed: 0,
  total: 0,
  unavailable: 0,
  groups: [],
  error: "",
};
export function hashDistance(a: string, b: string): number {
  if (!/^[0-9a-f]{16}$/i.test(a) || !/^[0-9a-f]{16}$/i.test(b)) return 65;
  let bits = BigInt(`0x${a}`) ^ BigInt(`0x${b}`),
    distance = 0;
  while (bits) {
    bits &= bits - 1n;
    distance++;
  }
  return distance;
}
export function buildPhotoGroups(
  photos: Photo[],
  fingerprints: PhotoFingerprint[],
  digests: Map<string, string> = new Map(),
): PhotoGroup[] {
  const byId = new Map(photos.map((p) => [p.id, p]));
  const prints = new Map(
    fingerprints.filter((p) => byId.has(p.id)).map((p) => [p.id, p]),
  );
  const quality = (id: string) => {
    const p = byId.get(id)!,
      f = prints.get(id);
    return (
      (p.favorite || f?.favorite ? 1e15 : 0) +
      p.width * p.height * 10 +
      (f?.quality || 0)
    );
  };
  const grouped = new Set<string>(),
    groups: PhotoGroup[] = [];
  const exact = new Map<string, string[]>();
  for (const [id, digest] of digests)
    if (byId.has(id) && /^[0-9a-f]{64}$/i.test(digest)) {
      const list = exact.get(digest) || [];
      list.push(id);
      exact.set(digest, list);
    }
  const add = (ids: string[], kind: PhotoGroup["kind"]) => {
    ids.sort((a, b) => quality(b) - quality(a) || a.localeCompare(b));
    ids.forEach((id) => grouped.add(id));
    groups.push({ id: `${kind}-${ids[0]}`, ids, recommended: ids[0]!, kind });
  };
  for (const ids of exact.values()) if (ids.length > 1) add(ids, "duplicate");
  // Compare with a group's representative, never chain unrelated images together.
  // Four 16-bit buckets bound comparisons on large libraries; timestamps limit false matches.
  const buckets = new Map<string, number[]>(),
    similar: string[][] = [];
  const ordered = [...photos].sort((a, b) => b.createdAt - a.createdAt);
  for (const photo of ordered) {
    const print = prints.get(photo.id);
    if (
      !print ||
      grouped.has(photo.id) ||
      photo.screenshot ||
      !/^[0-9a-f]{16}$/i.test(print.hash)
    )
      continue;
    const keys = Array.from(
      { length: 4 },
      (_, i) => `${i}:${print.hash.slice(i * 4, i * 4 + 4)}`,
    );
    const candidates = new Set(keys.flatMap((key) => buckets.get(key) || []));
    let match: number | undefined;
    for (const index of candidates) {
      const first = byId.get(similar[index]![0]!)!,
        firstPrint = prints.get(first.id)!;
      if (Math.abs(first.createdAt - photo.createdAt) > 86400000) continue;
      const ratio = first.width / first.height / (photo.width / photo.height);
      if (ratio < 0.9 || ratio > 1.1) continue;
      if (hashDistance(firstPrint.hash, print.hash) <= 7) {
        match = index;
        break;
      }
    }
    if (match !== undefined) similar[match]!.push(photo.id);
    else {
      const index = similar.push([photo.id]) - 1;
      keys.forEach((key) => {
        const list = buckets.get(key) || [];
        list.push(index);
        buckets.set(key, list);
      });
    }
  }
  similar.filter((ids) => ids.length > 1).forEach((ids) => add(ids, "similar"));
  return groups;
}
export function duplicateCandidates(prints: PhotoFingerprint[]): string[] {
  const byHash = new Map<string, string[]>();
  for (const p of prints)
    if (p.exactEligible) {
      const ids = byHash.get(p.hash) || [];
      ids.push(p.id);
      byHash.set(p.hash, ids);
    }
  return [...byHash.values()].filter((ids) => ids.length > 1).flat();
}

File: native/PhotoSweep/Services/Analysis.swift
import Foundation

struct PhotoGroup: Identifiable { var id: String; var ids: [String]; var recommended: String; var kind: String }
actor AnalysisService {
    func groups(_ photos: [MediaItem]) -> [PhotoGroup] {
        let photos = photos.filter { $0.kind == .photo }
        var prints: [String: [String: Any]] = [:]
        for start in stride(from: 0, to: photos.count, by: 24) {
            if Task.isCancelled { return [] }
            for value in PhotoSweepAnalysis.fingerprints(Array(photos[start..<min(start + 24, photos.count)]).map(\.id)) {
                if let id = value["id"] as? String { prints[id] = value }
            }
        }
        let byID = Dictionary(uniqueKeysWithValues: photos.map { ($0.id, $0) })
        func quality(_ id: String) -> Double {
            guard let photo = byID[id] else { return 0 }
            return (photo.favorite ? 1e15 : 0) + Double(photo.width * photo.height) * 10 + (prints[id]?["quality"] as? Double ?? 0)
        }
        var hashes: [String: [String]] = [:], digests: [String: [String]] = [:]
        for (id, value) in prints where value["exactEligible"] as? Bool == true { if let hash = value["hash"] as? String { hashes[hash, default: []].append(id) } }
        let potential = hashes.values.filter { $0.count > 1 }.flatMap { $0 }
        for start in stride(from: 0, to: potential.count, by: 8) {
            if Task.isCancelled { return [] }
            for value in PhotoSweepAnalysis.digests(Array(potential[start..<min(start + 8, potential.count)])) {
                if let id = value["id"], let digest = value["digest"] { digests[digest, default: []].append(id) }
            }
        }
        var groups: [PhotoGroup] = [], grouped = Set<String>()
        func add(_ ids: [String], kind: String) {
            let sorted = ids.sorted { quality($0) == quality($1) ? $0 < $1 : quality($0) > quality($1) }
            groups.append(PhotoGroup(id: kind + sorted[0], ids: sorted, recommended: sorted[0], kind: kind)); grouped.formUnion(ids)
        }
        for ids in digests.values where ids.count > 1 { add(ids, kind: "duplicate") }
        var similar: [[String]] = [], buckets: [String: [Int]] = [:]
        for photo in photos.sorted(by: { $0.createdAt > $1.createdAt }) {
            guard !grouped.contains(photo.id), !photo.screenshot, let hash = prints[photo.id]?["hash"] as? String, hash.count == 16, let number = UInt64(hash, radix: 16) else { continue }
            let keys = (0..<4).map { "\($0):\((number >> ($0 * 16)) & 0xffff)" }
            let candidates = Set(keys.flatMap { buckets[$0] ?? [] })
            let match = candidates.sorted().first { index in
                guard let first = byID[similar[index][0]], let firstHash = prints[first.id]?["hash"] as? String, let firstNumber = UInt64(firstHash, radix: 16), first.height > 0, photo.height > 0 else { return false }
                let ratio = Double(first.width) / Double(first.height) / (Double(photo.width) / Double(photo.height))
                return abs(first.createdAt - photo.createdAt) <= 86_400_000 && (0.9...1.1).contains(ratio) && (number ^ firstNumber).nonzeroBitCount <= 7
            }
            if let match { similar[match].append(photo.id) } else {
                let index = similar.count; similar.append([photo.id]); for key in keys { buckets[key, default: []].append(index) }
            }
        }
        for ids in similar where ids.count > 1 { add(ids, kind: "similar") }
        return groups
    }
}
